sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala

```diff
@@ -1325,6 +1325,13 @@ object RapidsConf {
     .booleanConf
     .createWithDefault(value = false)
 
+  val MAX_CONDITION_BRANCH_NUMBER = conf("spark.rapids.sql.opt.condition.maxBranchNumber")
+    .doc("Maximum number of branches for GPU case-when to enable the lazy evaluation of true " +
+      "and else expressions if the predicates on a batch are all-true or all-false. Big number " +
+      "may get GPU OOM easily since the predicates are cached during the computation.")
+    .integerConf
+    .createWithDefault(2)
```

Comment:
I'm not a fan of this config.  Users will not know what to set it to for their queries, nor will they realize this config is what needs to be tuned if the GPU runs out of memory due to it.

For small input batches we probably could cache far more than 2, but for very large batches we may not even want to do 2.  What we really want is a budget, in memory size not branch count or row sizes, that the CaseWhen expression is allowed to use.  Unfortunately we don't currently have a clear budget for this (and other expressions could similarly leverage this budget for its processing/optimizations).

cc: @revans2 since he may have further thoughts on this.

Reply:
Agree. 
Besides, different GPUs have different memory sizes, a value can work on A100 but may not work on T4.
For most cases, I think, users do not need to take care of this config, since 2 should be small enough to run without GPU OOM.